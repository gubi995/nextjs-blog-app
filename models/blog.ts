export interface Blog {
  id: string;
  title: string;
  createdAt: string;
  introduction: string;
  content: string;
  slug: string;
  image: {
    url: string;
  };
}
