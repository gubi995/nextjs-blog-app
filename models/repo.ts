export interface Repo {
  createdAt: string;
  name: string;
  homepageUrl: string;
  url: string;
}
