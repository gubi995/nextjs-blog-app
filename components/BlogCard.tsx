import React from 'react';
import { Card, CardContent, Typography, CardActions, Button, styled, CardMedia, Chip } from '@material-ui/core';

import { Blog } from '../models/blog';
import { toReadableDate } from '../utils/date';

const StyledCard = styled(Card)(({ theme }) => ({ maxWidth: 345, margin: theme.spacing(2) }));

const StyledCardMedia = styled(CardMedia)({ height: 140 });

const StyledCardActions = styled(CardActions)({ display: 'flex', justifyContent: 'center' });

const StyledChip = styled(Chip)(({ theme }) => ({ marginBottom: theme.spacing(1) }));

interface Props {
  blog: Blog;
  onClick: () => void;
}

const BlogCard = ({ blog, onClick }: Props) => {
  return (
    <StyledCard>
      <StyledCardMedia image={blog.image.url} />
      <CardContent>
        <Typography color="primary" gutterBottom variant="h5" component="h2">
          {blog.title}
        </Typography>
        <StyledChip color="primary" size="small" label={toReadableDate(blog.createdAt)} variant="outlined" />
        <Typography color="primary" variant="body2" component="p">
          {blog.introduction}
        </Typography>
      </CardContent>
      <StyledCardActions>
        <Button variant="outlined" color="primary" onClick={onClick}>
          View post
        </Button>
      </StyledCardActions>
    </StyledCard>
  );
};

export default BlogCard;
