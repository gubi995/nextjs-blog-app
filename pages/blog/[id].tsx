import { NextPage, GetStaticProps, GetStaticPaths } from 'next';
import { request } from 'graphql-request';
import { Container, styled, Typography } from '@material-ui/core';

import { Blog } from '../../models/blog';
import { toReadableDate } from '../../utils/date';

import AppShell from '../../components/AppShell';
import { GET_BLOG, GET_BLOG_IDS } from '../../graphql/queries';

const StyledImg = styled('img')(({ theme }) => ({
  marginTop: theme.spacing(3),
  marginBottom: theme.spacing(1),
  borderRadius: theme.shape.borderRadius,
  width: '100%',
  maxHeight: '400px',
}));

interface Props {
  blog: Blog;
}

const SingleBlogPostPage: NextPage<Props> = ({ blog }) => {
  return (
    <AppShell pageTitle={blog.title}>
      <Container maxWidth="sm">
        <StyledImg src={blog.image.url} />
        <Typography color="primary" variant="h3">
          {blog.title}
        </Typography>
        <Typography color="primary" variant="subtitle1">
          {toReadableDate(blog.createdAt)}
        </Typography>
        <div dangerouslySetInnerHTML={{ __html: blog.content.html }} />
      </Container>
    </AppShell>
  );
};

export const getStaticProps: GetStaticProps = async ({ params }) => {
  const data = await request(process.env.API_ENDPOINT!, GET_BLOG, { id: params?.id });

  const blog = data.blog;

  return {
    props: { blog },
  };
};

export const getStaticPaths: GetStaticPaths = async () => {
  const data = await request(process.env.API_ENDPOINT!, GET_BLOG_IDS);

  const paths = data.blogs.map((blog: Blog) => ({ params: { id: blog.id } }));

  return {
    paths,
    fallback: false,
  };
};

export default SingleBlogPostPage;
