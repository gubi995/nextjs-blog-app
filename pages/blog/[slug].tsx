import { NextPage, GetStaticProps, GetStaticPaths } from 'next';
import { Container, styled, Typography } from '@material-ui/core';

import { Blog } from '../../models/blog';
import { toReadableDate } from '../../utils/date';
import client from '../../utils/graphql-client';

import AppShell from '../../components/AppShell';
import { GET_BLOG, GET_BLOG_SLUGS } from '../../graphql/queries';

const StyledImg = styled('img')(({ theme }) => ({
  marginTop: theme.spacing(3),
  marginBottom: theme.spacing(1),
  borderRadius: theme.shape.borderRadius,
  width: '100%',
  maxHeight: '400px',
}));

const StyledTypography = styled(Typography)({ fontWeight: 'bold' });

interface Props {
  blog: Blog;
}

const SingleBlogPostPage: NextPage<Props> = ({ blog }) => {
  return (
    <AppShell pageTitle={blog.title}>
      <Container maxWidth="md">
        <StyledImg src={blog.image.url} />
        <StyledTypography color="primary" variant="h4">
          {blog.title}
        </StyledTypography>
        <Typography color="primary" variant="caption">
          {toReadableDate(blog.createdAt)}
        </Typography>
        <div dangerouslySetInnerHTML={{ __html: blog.content }} />
      </Container>
    </AppShell>
  );
};

export const getStaticProps: GetStaticProps = async ({ params }) => {
  const data = await client.request(GET_BLOG, { slug: params?.slug });

  const blog = data.blog;

  return {
    props: { blog },
  };
};

export const getStaticPaths: GetStaticPaths = async () => {
  const data = await client.request(GET_BLOG_SLUGS);

  const paths = data.allBlogs.map((blog: Blog) => ({ params: { slug: blog.slug } }));

  return {
    paths,
    fallback: false,
  };
};

export default SingleBlogPostPage;
