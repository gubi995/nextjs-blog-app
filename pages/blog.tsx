import { NextPage, GetStaticProps } from 'next';
import { Container, styled } from '@material-ui/core';
import { request } from 'graphql-request';

import { GET_BLOGS } from '../graphql/queries';
import { Blog } from '../models/blog';

import AppShell from '../components/AppShell';
import CustomBanner from '../components/CustomBanner';
import BlogCard from '../components/BlogCard';

const StyledContainer = styled(Container)({
  display: 'flex',
  flexWrap: 'wrap',
  justifyContent: 'center',
  alignItems: 'center',
});

interface Props {
  blogs: Blog[];
}

const BlogPage: NextPage<Props> = ({ blogs }) => {
  return (
    <AppShell pageTitle="Blog page">
      <CustomBanner title="Blog posts ✍" />
      <StyledContainer>
        {blogs.map((blog) => (
          <BlogCard blog={blog} key={blog.id} />
        ))}
      </StyledContainer>
    </AppShell>
  );
};

export async function getStaticProps(context: GetStaticProps) {
  const data = await request(process.env.API_ENDPOINT!, GET_BLOGS);

  const blogs = data.blogs;

  return {
    props: { blogs },
  };
}

export default BlogPage;
