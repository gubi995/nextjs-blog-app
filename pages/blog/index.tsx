import { NextPage, GetStaticProps } from 'next';
import Router from 'next/router';
import { Container, styled } from '@material-ui/core';
import { request } from 'graphql-request';

import { GET_BLOGS } from '../../graphql/queries';
import { Blog } from '../../models/blog';

import AppShell from '../../components/AppShell';
import CustomBanner from '../../components/CustomBanner';
import BlogCard from '../../components/BlogCard';

const StyledContainer = styled(Container)({
  display: 'flex',
  flexWrap: 'wrap',
  justifyContent: 'center',
  alignItems: 'flex-start',
});

interface Props {
  blogs: Blog[];
}

const BlogPostsPage: NextPage<Props> = ({ blogs }) => {
  const navigate = (id: string) => {
    Router.push('/blog/[id]', `/blog/${id}`);
  };

  return (
    <AppShell pageTitle="Blog page">
      <CustomBanner title="Blog posts ✍" />
      <StyledContainer>
        {blogs.map((blog) => (
          <BlogCard blog={blog} key={blog.id} onClick={() => navigate(blog.id)} />
        ))}
      </StyledContainer>
    </AppShell>
  );
};

export const getStaticProps: GetStaticProps = async (context) => {
  const data = await request(process.env.API_ENDPOINT!, GET_BLOGS);

  const blogs = data.blogs;

  return {
    props: { blogs },
  };
};

export default BlogPostsPage;
