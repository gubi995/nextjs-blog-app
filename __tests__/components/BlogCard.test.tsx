import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import BlogCard from '../../components/BlogCard';
import { Blog } from '../../models/blog';

const blog: Blog = {
  title: 'Blog title',
  content: 'Blog content',
  createdAt: '2020-04-01T07:14:32Z',
  id: '4489112',
  image: { url: 'https://www.mock.com/image.jpeg' },
  introduction: 'Blog intro',
  slug: 'blog-intro',
};

describe('BlogCard', () => {
  test("should rendered component matches with it's the snapshot", () => {
    const { asFragment } = render(<BlogCard blog={blog} onClick={jest.fn()} />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('should onClick has been called once', () => {
    const clickHandler = jest.fn();
    const { getByText } = render(<BlogCard blog={blog} onClick={clickHandler} />);

    const viewPostBtn = getByText('View post');

    fireEvent.click(viewPostBtn);

    expect(clickHandler).toHaveBeenCalledTimes(1);
  });
});
