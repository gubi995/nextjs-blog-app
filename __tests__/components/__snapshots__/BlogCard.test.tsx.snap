// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogCard should rendered component matches with it's the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root WithStyles(ForwardRef(Card))-root-1 WithStyles(ForwardRef(Card))-root-2 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardMedia-root WithStyles(ForwardRef(CardMedia))-root-32"
      style="background-image: url(https://www.mock.com/image.jpeg);"
    />
    <div
      class="MuiCardContent-root"
    >
      <h2
        class="MuiTypography-root MuiTypography-h5 MuiTypography-colorPrimary MuiTypography-gutterBottom"
      >
        Blog title
      </h2>
      <div
        class="MuiChip-root WithStyles(ForwardRef(Chip))-root-67 WithStyles(ForwardRef(Chip))-root-68 MuiChip-colorPrimary MuiChip-outlined MuiChip-outlinedPrimary MuiChip-sizeSmall"
      >
        <span
          class="MuiChip-label MuiChip-labelSmall"
        >
          4/1/2020
        </span>
      </div>
      <p
        class="MuiTypography-root MuiTypography-body2 MuiTypography-colorPrimary"
      >
        Blog intro
      </p>
    </div>
    <div
      class="MuiCardActions-root WithStyles(ForwardRef(CardActions))-root-99 MuiCardActions-spacing"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          View post
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
